microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java

```diff
@@ -51,84 +45,66 @@
-            MicroProfileHealthLogger.LOGGER.debugf("Discovered health check procedure %s", javaClass);
-            healthDelegates.add(event.getAnnotatedType());
-        } else if (event.getAnnotatedType().isAnnotationPresent(Liveness.class)) {
-            MicroProfileHealthLogger.LOGGER.debugf("Discovered liveness check procedure %s", javaClass);
-            livenessDelegates.add(event.getAnnotatedType());
-        } else if (event.getAnnotatedType().isAnnotationPresent(Readiness.class)) {
-            MicroProfileHealthLogger.LOGGER.debugf("Discovered liveness check procedure %s", javaClass);
-            readinessDelegates.add(event.getAnnotatedType());
-        }
     }
 
+    public CDIExtension(HealthReporter healthReporter) {
+        this.reporter = healthReporter;
+    }
 
     /**
-     * Instantiates <em>unmanaged instances</em> of HealthCheck and
-     * handle manually their CDI creation lifecycle.
-     * Add them to the {@link HealthReporter}.
+     * Get CDI <em>instances</em> of HealthCheck and
+     * add them to the {@link HealthReporter}.
      */
     private void afterDeploymentValidation(@Observes final AfterDeploymentValidation avd, BeanManager bm) {
-        addHealthCheck(bm, healthDelegates, healthInstances, reporter::addHealthCheck);
-        addHealthCheck(bm, readinessDelegates, readinessInstances, reporter::addReadinessCheck);
-        addHealthCheck(bm, livenessDelegates, livenessInstances, reporter::addLivenessCheck);
+        addHealthCheck(HealthLiteral.INSTANCE, bm, reporter::addHealthCheck);
+        addHealthCheck(LivenessLiteral.INSTANCE, bm, reporter::addLivenessCheck);
+        addHealthCheck(ReadinessLiteral.INSTANCE, bm, reporter::addReadinessCheck);
     }
 
-    private void addHealthCheck(BeanManager bm,
-                                List<AnnotatedType<? extends HealthCheck>> delegates,
-                                Collection<Unmanaged.UnmanagedInstance<HealthCheck>> instances,
+    private void addHealthCheck(AnnotationLiteral qualifier, BeanManager bm,
                                 Consumer<HealthCheck> healthFunction) {
-        for (AnnotatedType delegate : delegates) {
-            try {
-                Unmanaged<HealthCheck> unmanagedHealthCheck = new Unmanaged<HealthCheck>(bm, delegate.getJavaClass());
-                Unmanaged.UnmanagedInstance<HealthCheck> healthCheckInstance = unmanagedHealthCheck.newInstance();
-                HealthCheck healthCheck = healthCheckInstance.produce().inject().postConstruct().get();
-                instances.add(healthCheckInstance);
-                healthFunction.accept(healthCheck);
-            } catch (Exception e) {
-                throw new RuntimeException("Failed to register health bean", e);
-            }
+        Instance<HealthCheck> healthChecks = bm.createInstance().select(HealthCheck.class, qualifier);
+        for (HealthCheck healthCheck : healthChecks) {
+            healthFunction.accept(healthCheck);
```

Comment:
Can the bean instances obtained here be `@Dependent` beans?
If so then you are responsible for disposing them properly once you no longer need then in order to avoid leaks. CDI cannot control `@Dependent` bean's lifecycle if you retrieve them via `Instance` because normally the bean lifecycle is linked to the bean injecting it, but obviously, that won't work here.
You can destroy it with `Instance.destroy(healthCheck)` for every dependent bean you obtained (you can do that even for proxied beans but I am not sure that's desirable in this case?).